File: typescript/scripts/filter-vectors.mjs
/**
 * Regenerate `fixtures/cross-language/filter-vectors.json` from the reference implementation.
 *
 * The formatting layer is small and full of decisions a port can get subtly wrong: which end a
 * mask index counts from, whether a range may run backwards, whether an unknown filter throws or
 * passes the value through. Pinning the answers as data lets every implementation be asked the
 * same questions instead of each being argued with separately.
 */

import { writeFileSync } from 'node:fs';
import { dirname, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';

import { applyFilter } from '../dist/format/transforms.js';

const HERE = dirname(fileURLToPath(import.meta.url));
const OUT = resolve(HERE, '../../fixtures/cross-language/filter-vectors.json');

/** [filter, argument, input] — the argument is empty where the filter takes none. */
const CASES = [
  ['mask', 'xxx-xxx', '1234567'],
  ['mask', 'w[1] w[0]', 'John Smith'],
  ['mask', 'w[0] *', 'John Ronald Reuel Tolkien'],
  ['mask', 'x[0]. *', 'John Smith'],
  ['mask', '[tel.] xxx', '123'],
  ['mask', 'x[-1]', 'abcdef'],
  ['mask', 'x[1..3]', 'abcdef'],
  ['mask', 'x[3..1]', 'abcdef'],
  ['mask', 'x[9]', 'abc'],
  ['mask', '\\x xx', 'abc'],
  ['mask', 'w[-1], w[0]', 'Ann Bee Cee'],
  ['upper', '', 'straße'],
  ['lower', '', 'ÄÖÜ'],
  ['capitalize', '', 'мир труд'],
  ['title', '', 'mcDonald and sons'],
  ['slice', '1,4', 'abcdefg'],
  ['slice', '2', 'abcdefg'],
  ['slice', '-3', 'abcdefg'],
  ['replace', 'a,X', 'banana'],
  ['trim', '', '  padded  '],
  ['group', '3,-', '1234567'],
  ['group', '4, ', '1234567890'],
  ['compact', '', '1000000'],
  ['compact', '16', '255'],
  ['compact', '', 'not a number'],
  ['csv', '', 'Knife set, 3 "pcs"'],
  ['sql', '', "O'Brien"],
  ['nosuchfilter', '', 'left alone'],
];

const vectors = CASES.map(([kind, arg, input]) => ({
  kind,
  arg,
  input,
  expected: applyFilter(kind, arg === '' ? undefined : arg, input),
}));

const document = {
  schemaVersion: 1,
  $comment:
    'Interpolation filters and positional masks, as the TypeScript reference computes them. ' +
    'Regenerate with `npm run fixtures:filters`. The formatting layer is shared by three ' +
    'places that mean the same thing — the case= attribute, the compute tags and the ' +
    '${{Name|filter}} syntax — so one of these drifting shows up in all of them.',
  vectors,
};

writeFileSync(OUT, `${JSON.stringify(document, null, 2)}\n`);
console.error(`wrote ${OUT} (${String(vectors.length)} vectors)`);

File: rust/tests/filter_vectors.rs
//! The formatting layer, against the reference's own vectors.
//!
//! The layer is shared by three places that mean the same thing — the `case=`
//! attribute, the compute tags and the `${{Name|filter}}` syntax — so one of
//! them drifting shows up in all of them. These vectors pin the answer once,
//! which is why the fixture exists rather than each caller carrying its own
//! expectations.
//!
//! Note what an UNKNOWN filter does: it passes the value through untouched. That
//! is deliberate, and the validator is what complains about it (TDC192) — a
//! formatting layer that threw here would turn a typo into a crash halfway
//! through a run instead of a diagnostic before it starts.

mod common;

use tdcv2::format::{mask, transforms};
use tdcv2::json::Value;

fn apply(kind: &str, arg: &str, input: &str) -> String {
    if kind == "mask" {
        return mask::apply(arg, input).unwrap_or_else(|e| panic!("mask {arg:?}: {e}"));
    }
    if transforms::is_case_transform(kind) {
        return transforms::apply_case(kind, input);
    }
    let argument = (!arg.is_empty()).then_some(arg);
    transforms::apply_filter(kind, argument, input)
        .unwrap_or_else(|e| panic!("{kind}({arg:?}): {e}"))
}

#[test]
fn every_filter_vector_matches_the_reference() {
    let fixture = common::read_fixture("filter-vectors.json");
    let vectors = fixture
        .get("vectors")
        .and_then(Value::as_array)
        .expect("the fixture has vectors");
    assert!(vectors.len() >= 25, "only {} vectors", vectors.len());

    let text = |node: &Value, key: &str| {
        node.get(key)
            .and_then(Value::as_str)
            .unwrap_or_default()
            .to_string()
    };

    let mut checked = 0;
    for vector in vectors {
        let (kind, arg, input, expected) = (
            text(vector, "kind"),
            text(vector, "arg"),
            text(vector, "input"),
            text(vector, "expected"),
        );
        assert_eq!(
            apply(&kind, &arg, &input),
            expected,
            "{kind}({arg:?}) over {input:?}"
        );
        checked += 1;
    }
    println!("filter vectors: {checked} match the reference");
}
